src/background/Wallet/Wallet.ts

```diff
@@ -1536,20 +1536,42 @@ class PublicController {
   async wallet_switchEthereumChain({
     params,
     context,
+    id,
   }: PublicMethodParams<[{ chainId: string | number }]>): Promise<
     null | object
   > {
     const currentAddress = this.wallet.readCurrentAddress();
-    if (!currentAddress) {
-      throw new Error('Wallet is not initialized');
-    }
-    if (!this.wallet.allowedOrigin(context, currentAddress)) {
+    if (!context || !context.origin) {
       throw new OriginNotAllowed();
     }
     invariant(params[0], () => new InvalidParams());
     const { origin } = context;
     const { chainId: chainIdParameter } = params[0];
     const chainId = ethers.utils.hexValue(chainIdParameter);
+
+    if (
+      !currentAddress ||
+      !this.wallet.allowedOrigin(context, currentAddress)
+    ) {
+      const networks = await networksStore.load();
+      const chain = networks.getChainById(chainId);
+      return new Promise((resolve, reject) => {
+        this.safeOpenDialogWindow(origin, {
+          requestId: `${context.origin}:${id}`,
+          route: '/switchEthereumChain',
+          search: `?origin=${origin}&chainId=${chainId}`,
+          onResolve: () => {
+            this.wallet.setChainForOrigin(chain, origin);
+            resolve(null);
```

Comment:
In the default flow, we resolve in the next tick after updating selected chain:
```
      this.wallet.setChainForOrigin(chain, origin);
      // return null in next tick to give provider enough time to change chainId property
      return new Promise((resolve) => {
        setTimeout(() => resolve(null));
      });
```

Is there a reason not to do the same here?